### PR TITLE
Open document tree

### PR DESCRIPTION
**Pull request by tobiasKaminsky**
TODO
- [ ] same approach for public Cursor queryChildDocuments(String parentDocumentId, String[] projection, String sortOrder)

To test
- use at least Android O
- use https://github.com/android/storage/tree/master/ActionOpenDocumentTree
- open NC via lower left buttom
- select directory
- see directory listing in android test app

### CODE REVIEW

File: src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.java [hunk 1]
```diff
@@ -96,6 +101,8 @@
     private FileDataStorageManager currentStorageManager;
     private Map<Long, FileDataStorageManager> rootIdToStorageManager;
     private OwnCloudClient client;
+    @Setter
+    private boolean isSyncRunning = false;
```
[the remark this answers is not in the document]
Comment (AndyScherzinger) [reply src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.java]: correct var name would also be simply `syncRunning`

**Comment by grote** [comment]
Thanks a lot for working on this!

On my personal test, I could upload one file now, but the second one failed with this log: 

```ruby
2019-07-23 11:17:54.478 D/ExistenceCheckRemoteOperation: Existence check for https://test.de/remote.php/webdav/backup/1/kv/%40pm%40/QG1ldGFA targeting for  existence finished with HTTP status 404(FAIL)
2019-07-23 11:17:54.497 V/UploadsStorageManager: Updating /storage/emulated/0/Android/media/com.nextcloud.android.beta/nextcloud/tmp/test1@test.de/QG1ldGFA with status:UPLOAD_IN_PROGRESS and result:UNKNOWN (old:/storage/emulated/0/Android/media/com.nextcloud.android.beta/nextcloud/tmp/test1@test.de/QG1ldGFA status:UPLOAD_IN_PROGRESS result:-1)
2019-07-23 11:17:54.497 V/UploadsStorageManager: Updating /storage/emulated/0/Android/media/com.nextcloud.android.beta/nextcloud/tmp/test1@test.de/QG1ldGFA with status=UPLOAD_IN_PROGRESS
2019-07-23 11:17:54.508 D/UploadsStorageManager: updateUpload returns with: 1 for file: /storage/emulated/0/Android/media/com.nextcloud.android.beta/nextcloud/tmp/test1@test.de/QG1ldGFA
2019-07-23 11:17:54.508 D/UploadsStorageManager: notifyObserversNow
2019-07-23 11:17:54.526 V/UploadsStorageManager: Updating /storage/emulated/0/Android/media/com.nextcloud.android.beta/nextcloud/tmp/test1@test.de/QG1ldGFA with status=UPLOAD_IN_PROGRESS
2019-07-23 11:17:54.534 D/UploadsStorageManager: updateUpload returns with: 1 for file: /storage/emulated/0/Android/media/com.nextcloud.android.beta/nextcloud/tmp/test1@test.de/QG1ldGFA
2019-07-23 11:17:54.534 D/UploadsStorageManager: notifyObserversNow
2019-07-23 11:17:54.537 D/OwnCloudClient #6: REQUEST PUT /remote.php/webdav/backup/1/kv/%40pm%40/QG1ldGFA
2019-07-23 11:17:54.662 I/ReadFolderRemoteOperation: Synchronized /backup/1/kv/@pm@/: Operation finished with HTTP status code 207 (success)
2019-07-23 11:17:54.663 D/RefreshFolderOperation: Synchronizing test1@test.de/backup/1/kv/@pm@/
2019-07-23 11:17:54.675 D/RefreshFolderOperation: Remote folder /backup/1/kv/@pm@/ changed - starting update of local data 
2019-07-23 11:17:54.711 D/FileDataStorageManager: Saving folder /backup/1/kv/@pm@/ with 1 children and 0 files to remove
2019-07-23 11:17:54.721 D/FileDataStorageManager: Sending 2 operations to FileContentProvider
2019-07-23 11:17:54.722 D/FileContentProvider: applying batch in provider com.owncloud.android.providers.FileContentProvider@2e349a7 (temporary: false)
2019-07-23 11:17:54.727 D/FileContentProvider: applied batch in provider com.owncloud.android.providers.FileContentProvider@2e349a7
2019-07-23 11:17:54.739 D/RefreshFolderOperation: Send broadcast com.owncloud.android.operations.RefreshFolderOperation.EVENT_SINGLE_FOLDER_CONTENTS_SYNCED
2019-07-23 11:17:54.741 D/RefreshFolderOperation: Send broadcast com.owncloud.android.operations.RefreshFolderOperation.EVENT_SINGLE_FOLDER_SHARES_SYNCED
2019-07-23 11:17:54.747 E/DatabaseUtils: Writing exception to parcel
    java.lang.NullPointerException: Attempt to invoke virtual method 'long com.owncloud.android.datamodel.OCFile.getFileId()' on a null object reference
        at com.owncloud.android.providers.DocumentsStorageProvider.createFile(DocumentsStorageProvider.java:615)
        at com.owncloud.android.providers.DocumentsStorageProvider.createDocument(DocumentsStorageProvider.java:542)
        at android.provider.DocumentsProvider.callUnchecked(DocumentsProvider.java:1005)
        at android.provider.DocumentsProvider.call(DocumentsProvider.java:951)
        at android.content.ContentProvider$Transport.call(ContentProvider.java:404)
        at android.content.ContentProviderNative.onTransact(ContentProviderNative.java:272)
        at android.os.Binder.execTransact(Binder.java:731)
2019-07-23 11:17:54.748 W/DocumentsContract: Failed to create document
    java.lang.NullPointerException: Attempt to invoke virtual method 'long com.owncloud.android.datamodel.OCFile.getFileId()' on a null object reference
        at android.os.Parcel.createException(Parcel.java:1956)
        at android.os.Parcel.readException(Parcel.java:1918)
        at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:183)
        at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:135)
        at android.content.ContentProviderProxy.call(ContentProviderNative.java:651)
        at android.content.ContentProviderClient.call(ContentProviderClient.java:492)
        at android.provider.DocumentsContract.createDocument(DocumentsContract.java:1171)
        at android.provider.DocumentsContract.createDocument(DocumentsContract.java:1153)
        at androidx.documentfile.provider.TreeDocumentFile.createFile(TreeDocumentFile.java:53)
        at androidx.documentfile.provider.TreeDocumentFile.createFile(TreeDocumentFile.java:45)
```

**Comment by tobiasKaminsky** [comment]
> On my personal test, I could upload one file now, but the second one failed with this log:

How do you upload? 
I tried it with documentsui/files app found on emulator:

- go to images
- select three images
- select copy to
- go to nc, into a subfolder
- select create

--> I could create three images.

**Comment by grote** [comment]
I am using AndroidX's `DocumentFile` interface, but I could just reproduce it as well with the first three images I copied following your steps:

```ruby
2019-07-23 14:13:35.833 E/DatabaseUtils: Writing exception to parcel
    java.lang.NullPointerException: Attempt to invoke virtual method 'long com.owncloud.android.datamodel.OCFile.getFileId()' on a null object reference
        at com.owncloud.android.providers.DocumentsStorageProvider.createFile(DocumentsStorageProvider.java:615)
        at com.owncloud.android.providers.DocumentsStorageProvider.createDocument(DocumentsStorageProvider.java:542)
        at android.provider.DocumentsProvider.callUnchecked(DocumentsProvider.java:1005)
        at android.provider.DocumentsProvider.call(DocumentsProvider.java:951)
        at android.content.ContentProvider$Transport.call(ContentProvider.java:404)
        at android.content.ContentProviderNative.onTransact(ContentProviderNative.java:272)
        at android.os.Binder.execTransact(Binder.java:731)
2019-07-23 14:13:35.834 D/Metrics: docsui_other_fileop_failure: 5
2019-07-23 14:13:35.835 E/CopyJob: Failed to copy content://com.android.providers.downloads.documents/document/4
    com.android.documentsui.services.ResourceException: Couldn't create destination document 320px-RedCat_8727.jpg in directory content://org.nextcloud.beta.documents/document/14 due to an exception.
        at com.android.documentsui.services.CopyJob.byteCopyDocument(CopyJob.java:446)
        at com.android.documentsui.services.CopyJob.processDocument(CopyJob.java:399)
        at com.android.documentsui.services.CopyJob.start(CopyJob.java:287)
        at com.android.documentsui.services.Job.run(Job.java:156)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:458)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'long com.owncloud.android.datamodel.OCFile.getFileId()' on a null object reference
        at android.os.Parcel.createException(Parcel.java:1956)
        at android.os.Parcel.readException(Parcel.java:1918)
        at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:183)
        at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:135)
        at android.content.ContentProviderProxy.call(ContentProviderNative.java:651)
        at android.content.ContentProviderClient.call(ContentProviderClient.java:492)
        at android.provider.DocumentsContract.createDocument(DocumentsContract.java:1171)
        at com.android.documentsui.services.CopyJob.byteCopyDocument(CopyJob.java:441)
        at com.android.documentsui.services.CopyJob.processDocument(CopyJob.java:399) 
        at com.android.documentsui.services.CopyJob.start(CopyJob.java:287) 
        at com.android.documentsui.services.Job.run(Job.java:156) 
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:458) 
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:764) 
```

Two images were copied, but one threw this exception and an error dialog was shown.

**Comment by grote** [comment]
Something else I noticed: Folders created with Android's file manager don't show up immediately, just after a refresh. Not sure if that is the expected behavior or not.

Also, removing a file causes a `java.lang.UnsupportedOperationException: Remove not supported`. That can probably get added in a dedicated PR.

**Comment by tobiasKaminsky** [comment]
> Something else I noticed: Folders created with Android's file manager don't show up immediately, just after a refresh. Not sure if that is the expected behavior or not.

I assume this is a general limitation as it also happen on other providers.
(correct me if I am wrong)



> Also, removing a file causes a `java.lang.UnsupportedOperationException: Remove not supported`. That can probably get added in a dedicated PR.

Oh…indeed this is not yet implemented.
I'll open up a new PR for this.

**Comment by tobiasKaminsky** [comment]
> I am using AndroidX's `DocumentFile` interface, but I could just reproduce it as well with the first three images I copied following your steps:

This is strange. From where did you copied these files?

**Comment by grote** [comment]
> This is strange. From where did you copied these files?

I copied them from the Downloads folder on internal storage.

Just tried again. At least here on an API 28 phone, I can reproduce this behavior when copying three files. However, it is not always the same file that fails. Also saw 2 out of 3 failing.

The exception is always thrown from `com.owncloud.android.providers.DocumentsStorageProvider.createFile(DocumentsStorageProvider.java:615)`.

**Comment by jmue** [comment]
> Something else I noticed: Folders created with Android's file manager don't show up immediately, just after a refresh. Not sure if that is the expected behavior or not.
> 
> Also, removing a file causes a `java.lang.UnsupportedOperationException: Remove not supported`. That can probably get added in a dedicated PR.

To refresh folder contents after create/rename/copy/move operations you also have to notify the client. A `context.getContentResolver().notifyChange(toNotifyUri(document.getParent()), null, false);` is sufficient here (see my PR).

**Comment by tobiasKaminsky** [comment]
close in favor of #3905

**Pull request closed without merging**